feat(insights): Parameterized first day of the week in queries

Twixes opened the pull request:
## Problem

All our weeks start on Sunday, which is annoyingly US-centric. See ZEN-3780, ZEN-3700, ZEN-3660, ZEN-522, ZEN-3849, ZEN-3833

## Changes

Now in Project Settings:

<img width="342" alt="Screenshot 2023-06-23 at 17 42 40" src="https://github.com/PostHog/posthog/assets/4550621/e131122c-fb36-4473-93be-9923b7a4b77c">

## How did you test this code?

TODO

Twixes committed:
feat(insights): Allow either Monday or Sunday as week start

Twixes committed:
"Weeks start on" to "Week starts on"

Twixes committed:
Add `WeekStartDay` choices class

Twixes committed:
Wrap interval SQL in `toDateTime()`

Twixes committed:
Add printer test

Twixes committed:
Fix malformed SQL

Twixes committed:
Fix `BREAKDOWN_QUERY_SQL`

Twixes committed:
Fix excess parenthesis

Twixes committed:
Update test_event_query.py

Twixes committed:
Fix `NULL_SQL`

Twixes committed:
Add `team` to `Filter()` in a bunch of places

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Fix some num queries assertions

Twixes committed:
Fix some timezone issues

Twixes committed:
Fix funnel trends query syntax

Twixes committed:
Fix a few issues

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

Twixes committed:
Merge branch 'master' into week-start-setting

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

raquelmsmith commented:
Got another request for this https://posthogusers.slack.com/archives/C01GLBKHKQT/p1689255376235759

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Merge branch 'master' into week-start-setting

github-actions[bot] committed:
Update query snapshots

Twixes committed:
Fix some query issues

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Twixes committed:
Fix properties timeline margin

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
De-HogQL-ify `get_start_of_interval_sql`

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Make `get_start_of_interval_sql` robust

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Clean up and fix a missing timezone

Twixes committed:
Restore snapshots from master

Twixes committed:
Make date and time settings nice

thmsobrmlr commented on posthog/api/test/__snapshots__/test_insight_funnels.ambr:
```diff
@@ -1,408 +0,0 @@
-# name: ClickhouseTestFunnelTypes.test_funnel_time_to_convert_auto_bins
```
Why is this snapshot missing? The test still seems to be decorated with `@snapshot_clickhouse_queries`.

Twixes replied:
Looks like it was some quirk of tests, the snapshot is not changed anymore

thmsobrmlr commented on posthog/hogql/database/database.py:
```diff
@@ -92,12 +100,15 @@ def add_warehouse_tables(self, **field_definitions: Any):
             setattr(self, f_name, f_def)
 
 
+BUILT_IN_TABLE_NAMES = [attr for attr in dir(Database) if isinstance(getattr(Database, attr), Table)]
```
This change to table names seems unrelated to the week start date and I don't really understand what it's for - maybe separate this out into a separate PR.

Twixes replied:
Yes, definitely. It was also leftover from a previous HogQL-based approach

thmsobrmlr reviewed (commented):
Thanks for tackling this rabbit hole, @Twixes. Definitely needed for our European customers.

The PR was quite large with a number of loosely related changes. I've done my best to review this and left a couple of comments inline. For a ✅ , I'd love if you could:
- Go over the comments to check if something is suspicious (I've left a couple of questions where I was unsure).
- Check the two or three issues mentioned in comments as well.
- Move the changes to HogQL table names to a separate PR or help me understand it.
- It might be good to add a test for each of the insights with start of week = Monday.

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Fix SUNDAY/MONDAY

Co-authored-by: Thomas Obermüller <thomas.obermueller@gmail.com>

Twixes commented on posthog/api/test/test_insight.py:
```diff
@@ -1248,7 +1248,7 @@ def test_insight_trends_breakdown_pagination(self) -> None:
                 },
             )
-        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.json())
```
Just a debugging aid, previously if the response was not a 200 an extra `print()` was necessary to find out _why_ it wasn't a 200. With this, the response body is automatically shown in such case

Twixes committed:
Revert `BUILT_IN_TABLE_NAMES`

Twixes committed:
Revert new UI

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

Twixes commented on frontend/src/scenes/project/Settings/WeekStartConfig.tsx:
```diff
@@ -0,0 +1,35 @@
+    const { currentTeam, currentTeamLoading } = useValues(teamLogic)
+    const { updateCurrentTeam } = useActions(teamLogic)
+
+    const isTeamKnown = isAuthenticatedTeam(currentTeam)
```
Nope, it's really just about satisfying TypeScript. But looks like this isn't actually needed in this case, removed in #17231

Twixes committed:
Fix `_table_names`

Twixes commented on posthog/test/test_feature_flag.py:
```diff
@@ -2035,7 +2035,7 @@ def test_cohort_cache_no_unnecessary_queries(self):
         Person.objects.create(team=self.team, distinct_ids=["example_id"], properties={"email": "tim@posthog.com"})
 
-        with self.assertNumQueries(6):
+        with self.assertNumQueries(16):
```
Very good point, this was because the previous HogQL-based approach required all `Filter`'s to have `team`, which for some reason causes the persons-on-events instance setting to be selected a lot. For now just rolled back the `cohort.py` change

Twixes committed:
Revert change adding more queries

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Fix minor issues flagged by tests

github-actions[bot] committed:
Update query snapshots

Twixes committed:
Merge branch 'master' into week-start-setting

Twixes committed:
Split up `test_interval_filtering`

Twixes committed:
Test Monday/Sunday setting in `test_timezones_weekly`

Twixes commented:
Thanks for the great points @thmsobrmlr! They should be addressed here now, but take a good look. To make this change more tolerable, I split out the UI change into #17231.
As for tests, I added a test case for this feature within [test_timezones_weekly](https://github.com/PostHog/posthog/pull/16229/commits/eccfcdac5c9ca271d876c9a7761499d6cddb5106). That's not particularly extensive, but it's the gist of it

github-actions[bot] committed:
Update query snapshots

thmsobrmlr commented on ee/clickhouse/queries/test/__snapshots__/test_event_query.ambr:
```diff
@@ -74,24 +74,18 @@
-                   AND ((has(['test'], replaceRegexpAll(JSONExtractRaw(properties, 'name'), '^"|"$', '')))) )
-            GROUP BY id
-            HAVING max(is_deleted) = 0
-            AND ((has(['test'], replaceRegexpAll(JSONExtractRaw(argMax(person.properties, version), 'name'), '^"|"$', ''))))))
```
Seems correct, but I can't figure out where this change is coming from.

Twixes replied:
I don't know either 🤔 Even though it looks good

Twixes replied:
Oh, that's because of this diff https://github.com/PostHog/posthog/pull/16229/files#diff-81b09b4d64b61fe9e989ab1e49f40a4a7885d38e41df15eac395715da1dc3371
Also leftover from the previous HogQL-based approach, reverted

thmsobrmlr reviewed (approved):
Looks good!

Twixes committed:
Add `WeekStartDay` cast

Twixes commented on ee/clickhouse/queries/test/__snapshots__/test_retention.ambr:
```diff
@@ -17,12 +17,12 @@
                  event_date),
           target_event_query as
-       (SELECT DISTINCT toStartOfWeek(toDateTime(e.timestamp), 0, 'UTC') AS event_date,
+       (SELECT DISTINCT toStartOfWeek(toTimeZone(toDateTime(e.timestamp, 'UTC'), 'UTC')) AS event_date,
```
It's clearer when [`@also_test_with_different_timezones` is used ](https://github.com/PostHog/posthog/pull/16229/files#diff-c3a80c29a590658f38d914aa588ca3a3c91379fe2b692728c75b95bec7717af7L5544)

Twixes committed:
Restore test_event_query.py

Twixes committed:
Merge branch 'master' into week-start-setting

thmsobrmlr reviewed (approved):
Looks good now!
